wordpress-mobile/WordPress-iOS
# Aztec: Implements basic state machine for editor state

Closes #6422 

This adds the Publish button (previously called Post) to the Aztec editor. Instead of using a series of hacky methods to determine the state of the button it extracts the logic into clearly defined units of state that is unit testable. Other variables that affect state of the editor are also in the state context.

This is partially based off of @frosty's PostEditorSaveAction enum and WPPostViewController extension but made to be more abstract.

The main things this solves are:
* Publish button title.
* Publish button state (enabled/disabled).
* What text is shown during the publish process (Publishing...).

What it can/should solve but isn't part of this PR:
* Is Post-Post shown?
* Is there a secondary publish button and what does it say? (publish now, save draft, etc)
* Anything else that's related to the state of the editor that shouldn't turn into a pile of crazy If statements.

Some other TODOs that aren't part of this PR:
* **Submit for Review** is a lot of text to fit on the button label. Maybe just Submit? How are we handling it in cases where the button text is just too long in other languages?
* Submit for Review is also not currently supported in the hybrid editor. May need to validate how this all works. Currently the Aztec editor supports it but we're not setting the Bool properly if the user has or doesn't have the permission to publish.
* Update the state context when the post is being published to disable the publish button.
* Add some more unit tests to make sure all state switches are accounted for.

## How to Test
Make sure the publish button shows the right text and is enabled in these scenarios:
1. New post, no content. (Publish, disabled)
2. New post, with content. (Publish, enabled)
3. New post, set to draft. (Save)
4. New post, future dated. (Schedule)
5. New post, past dated. (Publish)
6. Existing published post. (Update)
7. Existing scheduled post. (Update)
8. Existing draft post. (Save)

There are a bunch of other scenarios but you get the point.

Needs review: @frosty, @jleandroperez
